src/lib/tpm.c

```diff
@@ -4685,3 +4685,29 @@ CK_RV tpm_ec_ecdh1_derive(tpm_ctx *tctx, tobject *tobj, uint8_t *ecc_point,
     return rv;
 }
 
+CK_RV tpm_get_esys_tr(tpm_ctx *ctx,
+                      uint32_t persistent_handle,
+                      uint32_t *esys_tr)
+{
+    ESYS_TR tr;
+    TSS2_RC rval;
+
+    if (!esys_tr) {
+        return CKR_GENERAL_ERROR;
+    }
+
+    rval = Esys_TR_FromTPMPublic(
+            ctx->esys_ctx,
+            persistent_handle,
+            ESYS_TR_NONE,
+            ESYS_TR_NONE,
+            ESYS_TR_NONE,
+            &tr);
```

Comment:
I missed this before, we want to use a session handle for session1 parameter for `Esys_TR_FromTPMPublic`. The handle to use is in `tpm_ctx `structure as field `hmac_session`, for example: `ctx->hmac_session`.

The reason that this is important, is this is where the attacker could try and swap out what the handle points to on the tpm as a way to get you use a key they supply. By using the session handle `hmac_session`, we are in an encrypted and integrity verified connection with the TPM. Think of it as ssh for the TPM.